spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpConnector.java

```diff
@@ -0,0 +1,169 @@
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.lang.Nullable;
+
+/**
+ * {@link ClientHttpConnector} implementation for the Apache HttpComponents HttpClient 5.x.
+ *
+ * @author Martin Tarjányi
+ * @since 5.3
+ * @see <a href="https://hc.apache.org/index.html">Apache HttpComponents</a>
+ */
+public class HttpComponentsClientHttpConnector implements ClientHttpConnector {
+
+	private final CloseableHttpAsyncClient client;
+
+	private final DataBufferFactory dataBufferFactory;
+
+	private final Supplier<? extends HttpClientContext> contextSupplier;
+
+
+	/**
+	 * Default constructor that creates and starts a new instance of {@link CloseableHttpAsyncClient}.
+	 */
+	public HttpComponentsClientHttpConnector() {
+		this(HttpAsyncClients.createDefault());
+	}
+
+	/**
+	 * Constructor with a pre-configured {@link CloseableHttpAsyncClient} instance.
+	 * @param client the client to use
+	 */
+	public HttpComponentsClientHttpConnector(CloseableHttpAsyncClient client) {
+		this(client, HttpClientContext::create);
+	}
+
+	/**
+	 * Constructor with a pre-configured {@link CloseableHttpAsyncClient} instance
+	 * and a {@link HttpClientContext} supplier lambda which is called before each request
+	 * and passed to the client.
+	 * @param client the client to use
+	 * @param contextSupplier a {@link HttpClientContext} supplier
+	 */
+	public HttpComponentsClientHttpConnector(CloseableHttpAsyncClient client,
+			Supplier<? extends HttpClientContext> contextSupplier) {
+
+		this.dataBufferFactory = new DefaultDataBufferFactory();
+		this.contextSupplier = contextSupplier;
+		this.client = client;
+		this.client.start();
+	}
+
+
+	@Override
+	public Mono<ClientHttpResponse> connect(HttpMethod method, URI uri,
+			Function<? super ClientHttpRequest, Mono<Void>> requestCallback) {
+
+		HttpClientContext context = this.contextSupplier.get();
+
+		if (context.getCookieStore() == null) {
+			context.setCookieStore(new BasicCookieStore());
+		}
+
+		HttpComponentsClientHttpRequest request = new HttpComponentsClientHttpRequest(method, uri,
+				context, this.dataBufferFactory);
+
+		return requestCallback.apply(request).then(Mono.defer(() -> execute(request, context)));
+	}
+
+	private Mono<ClientHttpResponse> execute(HttpComponentsClientHttpRequest request, HttpClientContext context) {
+		Flux<ByteBuffer> byteBufferFlux = request.getByteBufferFlux();
+
+		ReactiveEntityProducer reactiveEntityProducer = createReactiveEntityProducer(request, byteBufferFlux);
+
+		BasicRequestProducer basicRequestProducer = new BasicRequestProducer(request.getHttpRequest(),
+				reactiveEntityProducer);
+
+		return Mono.<Message<HttpResponse, Publisher<ByteBuffer>>>create(sink -> {
+			ReactiveResponseConsumer reactiveResponseConsumer =
+					new ReactiveResponseConsumer(new MonoFutureCallbackAdapter<>(sink));
+
+			this.client.execute(basicRequestProducer, reactiveResponseConsumer, context, null);
```

Comment:
Passing null here for the 4th parameter callback seems odd, but I suppose it is OK, because the completion signal is also passed to the response consumer, and from there to the callback adapter. Correct?

Reply:
Yes, that's my understanding.